src/error.rs

```diff
@@ -0,0 +1,154 @@
+    Unimplemented(NodePtr),
+
+    #[error("Operator Error: InvalidArg: {1}")]
+    InvalidArg(NodePtr, String),
+
+    #[error("CoinID Error: Invalid Parent Coin ID, not 32 bytes")]
+    CoinIDParentCoinIdNot32Bytes(NodePtr),
+
+    #[error("CoinID Error: Invalid Puzzle Hash, not 32 bytes")]
+    CoinIDPuzzleHashNot32Bytes(NodePtr),
+
+    #[error("CoinID Error: Invalid Amount: Amount is Negative")]
+    CoinIDAmountNegative(NodePtr),
+
+    #[error("CoinID Error: Invalid Amount: Amount has leading zeroes")]
+    CoinIDAmountLeadingZeroes(NodePtr),
+
+    #[error("CoinID Error: Invalid Amount: Amount exceeds max coin amount")]
+    CoinIDAmountExceedsMaxCoinAmount(NodePtr),
+
+    #[error("Allocator Error: {0}")]
+    Allocator(#[from] AllocatorErr),
+}
+impl From<std::io::Error> for EvalErr {
+    fn from(_: std::io::Error) -> Self {
+        EvalErr::SerializationError
+    }
+}
+
+impl EvalErr {
+    fn node(&self) -> Option<NodePtr> {
+        match self {
+            EvalErr::Raise(node) => Some(*node),
+            EvalErr::InvalidNilTerminator(node) => Some(*node),
+            EvalErr::DivisionByZero(node) => Some(*node),
+            EvalErr::DivmodByZero(node) => Some(*node),
+            EvalErr::ModByZero(node) => Some(*node),
+            EvalErr::ModPowZeroModulus(node) => Some(*node),
+            EvalErr::ShiftTooLarge(node) => Some(*node),
+            EvalErr::ValueStackLimitReached(node) => Some(*node),
+            EvalErr::EnvironmentStackLimitReached(node) => Some(*node),
+            EvalErr::InternalError(node, _) => Some(*node),
+            EvalErr::Reserved(node) => Some(*node),
+            EvalErr::Invalid(node) => Some(*node),
+            EvalErr::Unimplemented(node) => Some(*node),
+            EvalErr::CoinIDPuzzleHashNot32Bytes(node) => Some(*node),
+            EvalErr::CoinIDAmountNegative(node) => Some(*node),
+            EvalErr::CoinIDAmountLeadingZeroes(node) => Some(*node),
+            EvalErr::CoinIDAmountExceedsMaxCoinAmount(node) => Some(*node),
+            EvalErr::CoinIDParentCoinIdNot32Bytes(node) => Some(*node),
+            EvalErr::Allocator(alloc) => AllocatorErr::node(alloc),
+            EvalErr::InvalidArg(node, _) => Some(*node),
+            _ => None,
+        }
+    }
+    pub fn combined_str(&self) -> String {
+        // This is a convenience function to get the combined string representation of the error
+        match self.node() {
+            Some(node) => format!("{self}: {node:?}"),
+            None => self.to_string(),
+        }
+    }
+    pub fn node_ptr(&self) -> NodePtr {
+        // This is a convenience function to get the node pointer
+        self.node().unwrap_or_default()
+    }
+}
+
+impl PartialEq<Self> for EvalErr {
+    fn eq(&self, other: &Self) -> bool {
+        self.combined_str() == other.combined_str()
+    }
+}
+#[derive(Debug, Error)]
+pub enum AllocatorErr {
+    #[error("InvalidArg: {1}")]
+    InvalidArg(NodePtr, String),
+}
```

Comment:
the fact that there's only one error code for `AllocatorErr` makes me think that maybe it would be easier to just merge this with `EvalErr`. Or at least convert this error directly to `EvalErr::InvalidArg` rather than containing the allocator error.